core/domain/feature_flag_services.py

```diff
@@ -0,0 +1,282 @@
+                feature_flag_config_model.id] = (
+                    feature_flag_domain.FeatureFlag.from_dict({
+                        'name': feature_flag_config_model.id,
+                        'description': feature_flag_spec.description,
+                        'feature_stage': feature_flag_spec.feature_stage.value,
+                        'force_enable_for_all_users': (
+                            feature_flag_config_model.force_enable_for_all_users
+                        ),
+                        'rollout_percentage': (
+                            feature_flag_config_model.rollout_percentage),
+                        'user_group_ids': (
+                            feature_flag_config_model.user_group_ids),
+                        'last_updated': last_updated
+                    })
+                )
+
+        for feature_flag_name in feature_flag_names_list:
+            if feature_flag_name not in (
+                feature_flag_name_to_feature_flag_model_dict
+            ):
+                feature_flag_name_to_feature_flag_model_dict[
+                    feature_flag_name] = None
+
+    return feature_flag_name_to_feature_flag_model_dict
+
+
+def is_feature_flag_enabled(
+    user_id: Optional[str],
+    feature_flag_name: str,
+    feature_flag: Optional[feature_flag_domain.FeatureFlag] = None
+) -> bool:
+    """Returns True if feature is enabled for the given user else False.
+
+    Args:
+        user_id: str|None. The id of the user, if logged-out user then None.
+        feature_flag_name: str. The name of the feature flag that needs to
+            be evaluated.
+        feature_flag: FeatureFlag. The feature flag domain model.
+
+    Returns:
+        bool. True if the feature is enabled for the given user else False.
+    """
+    if feature_flag is None:
+        feature_flag = registry.Registry.get_feature_flag(feature_flag_name)
+
+    current_server = feature_flag_domain.get_server_mode()
+
+    if (
+        current_server == feature_flag_domain.ServerMode.TEST and
+        feature_flag.feature_flag_spec.feature_stage ==
+        feature_flag_domain.ServerMode.DEV
+    ):
+        return False
+
+    if (
+        current_server == feature_flag_domain.ServerMode.PROD and
+        feature_flag.feature_flag_spec.feature_stage in (
+            feature_flag_domain.ServerMode.DEV,
+            feature_flag_domain.ServerMode.TEST)
+    ):
+        return False
+
+    if feature_flag.feature_flag_config.force_enable_for_all_users:
+        return True
+    if user_id:
+        salt = feature_flag_name.encode('utf-8')
+        hashed_user_id = hashlib.sha256(
+            user_id.encode('utf-8') + salt).hexdigest()
+        hash_value = int(hashed_user_id, 16)
+        mod_result = hash_value % 1000
+        threshold = (
+            feature_flag.feature_flag_config.rollout_percentage / 100) * 1000
+        return bool(mod_result < threshold)
+    return False
+
+
+def evaluate_all_feature_flag_configs(
```

Comment:
In practice, what's the current runtime of this, if all flags have values in the datastore? (Just need to check that the time in practice is quite fast, since we'll do this all the time.)

Reply:
Around 77milliseconds